src/lib.rs

```diff
@@ -159,33 +159,34 @@ pub type LevelRun = Range<usize>;
 /// `line` is a range of bytes indices within `paragraph`.
 ///
 /// http://www.unicode.org/reports/tr9/#Reordering_Resolved_Levels
-pub fn visual_runs(line: Range<usize>,
-                   para_level: u8,
-                   max_level: u8,
-                   levels: &[u8]) -> Vec<LevelRun> {
-    assert!(line.start <= levels.len());
-    assert!(line.end <= levels.len());
+pub fn visual_runs(line: Range<usize>, info: &ParagraphInfo) -> Vec<LevelRun> {
+    assert!(line.start <= line.end);
+    assert!(line.end <= info.levels.len());
 
     // TODO: Whitespace handling.
     // http://www.unicode.org/reports/tr9/#L1
 
-    assert!(max_level >= para_level);
-    let mut runs = Vec::with_capacity((max_level - para_level) as usize + 1);
+    assert!(info.max_level >= info.para_level);
+    let mut runs = Vec::with_capacity((info.max_level - info.para_level) as usize + 1);
 
     // Optimization: If there's only one level, just return a single run for the whole line.
-    if max_level == para_level || line.len() == 0 {
+    if info.max_level == info.para_level || line.len() == 0 {
         runs.push(line.clone());
         return runs
     }
 
     // Find consecutive level runs.
     let mut start = line.start;
-    let mut level = levels[start];
-    let mut min_level = level;
-    let mut max_level = level;
+    let mut level = info.levels[start];
+    let mut min_level = level;  // FIXME: Is this the same as info.max_level?
+    let mut max_level = level;  // FIXME: Is this the same as info.para_level?
 
+    // FIXME: Could this loop be replaced with a call to level_runs()?
     for i in (start + 1)..line.end {
-        let new_level = levels[i];
+        if prepare::removed_by_x9(info.classes[i]) {
+            continue
```

Comment:
I don't think this is strictly necessary, since I believe it will only affect the ordering of removed characters, which is not defined by the spec (since they must be removed or rendered as zero-width anyway).  However, it should make layout more efficient by not splitting level runs needlessly.

I think a better way to do this would be to set each removed character's level to the level of the previous non-removed character, as in `assignLevelsToCharactersRemovedByX9` in the [reference implementation][1], since this will also help code like Servo that uses the returned `levels` array directly.

[1]: http://www.unicode.org/Public/PROGRAMS/BidiReferenceJava/BidiReference.java